### DISCUSSION TITLE
Using time scale for bar and line charts

### DISCUSSION THREAD

**Discussion by umarluqman** [airbnb/visx#1566] [Q&A]
<img width="527" alt="Screenshot 2022-09-13 at 8 57 24 AM" src="https://user-images.githubusercontent.com/25649761/189945957-063872e5-67f5-4905-b2c3-5c2dcf07f8a7.png">

I wonder is it possible to use time scale in x-axis like the image above? I knew it works well with the line but I'm not sure about using time scale for the bar.

**Comment by williaster** [comment]
Hey @umarluqman 👋 thanks for checking out `visx`. It is possible to use `time` scales for bars, but overall I would probably recommend using `band` scales.  

if you use a `time` scale, you have to manually compute the width of bars based on chart width and # of bars. this gets even more tricky if you have **missing** data (i.e., you don't have a data point for every date on the `time` scale), and if you want to have spaces/gaps between the bars. you can also end up with weird edge cases because time is complicated (e.g., daylight savings might shift the position of one bar slightly because of the extra hour or lost hour).

if you use a `band` scale, you get width and spacing for free. the only trick with these is that you need to fill in missing dates (this is pretty easy for `date`s, harder for `datetime`s) because a `band` scale is technically categorical so it views e.g.,. `2022-09-01` as a category not as a date.

**Comment by umarluqman** [comment]
Thanks for the reply @williaster , I would like to use the `band` scale but from my limited understanding, there is no invert function in the `band` scale to know the value in that exact tooltip location (given the user can hover anywhere in the graph and getting the tooltip). Do you know how do I achieve that?

**Comment by humanchimp** [comment]
i would also like to use a time scale for the axis and a band scale for the buckets themselves. currently, i am hand-coding this using the d3-scales and the `BarStack` and, yes, it is complicated.

**Comment by tkdgus96** [comment]
Using the `band` scale seems like a pretty good alternative. But in some cases this doesn't seem to be right.

The common factor in these cases are when the time interval is different for each bar. In other words, the problem happens when the spacing between each bar has to be different. For `BarStack` if we set the xScale as a timeScale, the width of each bar calculates to a abnormal value for some reason. In my case, I solved this problem by filling in the missing bars with a `null` value and calculating the width of each bar (dividing width by length).

Here is a sample code for the `BarStack` case

```typescript
const StackedBarChart: FC<Props> = ({
  width,
  height,
  data = mockData,
  colors = ['#6c5efb', '#c998ff', '#a44afe'],
}) => {
  const xScale = scaleTime({
    domain: extent(data.map((d) => new Date(d.date))) as [Date, Date],
    range: [0, width],
  });
  const yScale = scaleLinear<number>({
    domain: [
      0,
      Math.max(
        ...data
          .map((d) =>
            Object.values(d.data).reduce(
              (acc, curr) => Number(acc) + Number(curr),
              0,
            ),
          )
          .flat(),
      ),
    ],
    range: [height, 0],
  });
  const keys = Object.keys(data[0]?.data);
  const colorScale = scaleOrdinal<string, string>({
    domain: keys,
    range: colors,
  });
  return (
    <svg width={width} height={height}>
      <Group left={30}>
        <AxisLeft scale={yScale} />
        <AxisBottom scale={xScale} top={height} />
      </Group>
      <Group left={150}>
        <BarStack
          data={data.map((d) => ({ ...d.data, date: d.date }))}
          keys={keys}
          xScale={xScale}
          yScale={yScale}
          color={colorScale}
          x={(d) => new Date(d.date)}
        >
          {(barStacks) => {
            return barStacks.map((barStack) =>
              barStack.bars.map((bar) => (
                <rect
                  key={`bar-stack-${barStack.index}-${bar.index}`}
                  x={bar.x}
                  y={bar.y}
                  height={bar.height}
                  width={width / data.length}
                  fill={bar.color}
                />
              )),
            );
          }}
        </BarStack>
      </Group>
    </svg>
  );
};
```
